mailpoet/lib/Subscription/Captcha/Validator/BuiltInCaptchaValidator.php

```diff
@@ -0,0 +1,142 @@
+use MailPoet\Subscription\Captcha\CaptchaPhrase;
+use MailPoet\Subscription\Captcha\CaptchaSession;
+use MailPoet\Subscription\SubscriptionUrlFactory;
+use MailPoet\Util\Helpers;
+use MailPoet\WP\Functions as WPFunctions;
+
+class BuiltInCaptchaValidator implements CaptchaValidator {
+
+
+  /** @var SubscriptionUrlFactory  */
+  private $subscriptionUrlFactory;
+
+  /** @var CaptchaPhrase  */
+  private $captchaPhrase;
+
+  /** @var CaptchaSession  */
+  private $captchaSession;
+
+  /** @var WPFunctions  */
+  private $wp;
+
+  /** @var SubscriberIPsRepository  */
+  private $subscriberIPsRepository;
+
+  public function __construct(
+    SubscriptionUrlFactory $subscriptionUrlFactory,
+    CaptchaPhrase $captchaPhrase,
+    CaptchaSession $captchaSession,
+    WPFunctions $wp,
+    SubscriberIPsRepository $subscriberIPsRepository
+  ) {
+    $this->subscriptionUrlFactory = $subscriptionUrlFactory;
+    $this->captchaPhrase = $captchaPhrase;
+    $this->captchaSession = $captchaSession;
+    $this->wp = $wp;
+    $this->subscriberIPsRepository = $subscriberIPsRepository;
+  }
+
+  public function validate(array $data): bool {
+    $isBuiltinCaptchaRequired = $this->isRequired(isset($data['email']) ? $data['email'] : '');
+    if (!$isBuiltinCaptchaRequired) {
+      return true;
+    }
+    if (empty($data['captcha'])) {
+      throw new ValidationError(
+        $this->wp->__('Please fill in the CAPTCHA.', 'mailpoet'),
+        [
+          'redirect_url' => $this->subscriptionUrlFactory->getCaptchaUrl($this->getSessionId()),
+        ]
+      );
+    }
+    $captchaHash = $this->captchaPhrase->getPhrase();
+    if (empty($captchaHash)) {
+      throw new ValidationError(
+        $this->wp->__('Please regenerate the CAPTCHA.', 'mailpoet'),
+        [
+          'redirect_url' => $this->subscriptionUrlFactory->getCaptchaUrl($this->getSessionId()),
+        ]
+      );
+    }
+
+    if (!hash_equals(strtolower($data['captcha']), strtolower($captchaHash))) {
+      $this->captchaPhrase->resetPhrase();
+      throw new ValidationError(
+        $this->wp->__('The characters entered do not match with the previous CAPTCHA.', 'mailpoet'),
+        [
+          'refresh_captcha' => true,
+        ]
+      );
+    }
+
+    return true;
+
+  }
+
+  private function getSessionId() {
+    $id = $this->captchaSession->getId();
+    if ($id === null) {
+      $this->captchaSession->init();
+      $id = $this->captchaSession->getId();
+    }
+    return $id;
+  }
+
+  public function isRequired($subscriberEmail = null) {
+    if ($this->isUserExemptFromCaptcha()) {
+      return false;
+    }
+
+    $subscriptionCaptchaRecipientLimit = $this->wp->applyFilters('mailpoet_subscription_captcha_recipient_limit', 0);
+    if ($subscriptionCaptchaRecipientLimit === 0) {
+      return true;
+    }
+
+    // Check limits per recipient if enabled
+    if ($subscriberEmail) {
+      $subscriber = Subscriber::where('email', $subscriberEmail)->findOne();
```

Comment:
Why not `SubscriberEntity`? I think it's better to stick to Doctrine entities in new code so that we won't need to refactor Models out of them later on

Reply:
Thanks for that catch! Addressed in [c6cf795](https://github.com/mailpoet/mailpoet/pull/4196/commits/c6cf7956b072c15baf1dc1e61782b537a566e104)